### PR TITLE
Add performance measuring to join sort merge

### PR DESCRIPTION
**Pull request by Alexander-Dubrawski**
This PR allows us to analyze the performance of the `join sort merge` in more detail. Inside the `join sort merge` the `OperatorPerformanceData` class is used to measure the runtime of the different steps: 

* left side materializing
* right side materializing
* clustering
* merging
* output writing

Screenshot 2021-01-12 at 16 03 50

### CODE REVIEW

File: src/lib/operators/join_sort_merge/radix_cluster_sort.hpp [hunk 1]
```diff
@@ -325,20 +330,25 @@ class RadixClusterSort {
...
         left_column_materializer.materialize(_left_input_table, _left_column_id);
+    _performance.set_step_runtime(JoinSortMerge::OperatorSteps::LeftSideMaterializing, timer_materialization.lap());
     auto [materialized_right_segments, null_rows_right, samples_right] =
         right_column_materializer.materialize(_right_input_table, _right_column_id);
+    _performance.set_step_runtime(JoinSortMerge::OperatorSteps::RightSideMaterializing, timer_materialization.lap());
     output.null_rows_left = std::move(null_rows_left);
     output.null_rows_right = std::move(null_rows_right);
 
     // Append right samples to left samples and sort (reserve not necessarity when insert can
     // determined the new capacity from iterator: https://stackoverflow.com/a/35359472/1147726)
```
Comment (Bouncner) [inline src/lib/operators/join_sort_merge/radix_cluster_sort.hpp:347]: "determine"

File: src/lib/operators/join_sort_merge/radix_cluster_sort.hpp [hunk 2]
```diff
@@ -325,20 +330,25 @@ class RadixClusterSort {
...
     // Sort the chunks of the input tables in the non-equi cases
     ColumnMaterializer<T> left_column_materializer(!_equi_case, _materialize_null_left);
     ColumnMaterializer<T> right_column_materializer(!_equi_case, _materialize_null_right);
     auto [materialized_left_segments, null_rows_left, samples_left] =
         left_column_materializer.materialize(_left_input_table, _left_column_id);
+    _performance.set_step_runtime(JoinSortMerge::OperatorSteps::LeftSideMaterializing, timer_materialization.lap());
     auto [materialized_right_segments, null_rows_right, samples_right] =
         right_column_materializer.materialize(_right_input_table, _right_column_id);
+    _performance.set_step_runtime(JoinSortMerge::OperatorSteps::RightSideMaterializing, timer_materialization.lap());
     output.null_rows_left = std::move(null_rows_left);
```
Comment (Bouncner) [inline src/lib/operators/join_sort_merge/radix_cluster_sort.hpp:343]: Can you pull that null std::move directly behind the materialize side? So L343 after l338. Then we include the null move into the materialization side.

File: src/lib/operators/join_sort_merge.cpp [hunk 3]
```diff
@@ -922,12 +930,15 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
...
 
+    _performance.set_step_runtime(OperatorSteps::Merging, timer_merging.lap());
+
     // Intermediate structure for output chunks (to avoid concurrent appending to table)
     std::vector<std::shared_ptr<Chunk>> output_chunks(_output_pos_lists_left.size());
 
     // Threshold of expected rows per partition over which parallel output jobs are spawned.
     constexpr auto PARALLEL_OUTPUT_THRESHOLD = 10'000;
 
+    Timer timer_output_writing;
```
Comment (Bouncner) [inline src/lib/operators/join_sort_merge.cpp:941]: Should be fine to merge both timers here.

Comment (Bouncner) [reply src/lib/operators/join_sort_merge.cpp:941]: See the other comments.

File: src/lib/operators/join_sort_merge.cpp [hunk 4]
```diff
@@ -892,6 +898,8 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
     _end_of_left_table = _end_of_table(_sorted_left_table);
     _end_of_right_table = _end_of_table(_sorted_right_table);
 
+    Timer timer_merging;
```
Comment (Bouncner) [inline src/lib/operators/join_sort_merge.cpp:901]: Can we find a better name, now that the timer also includes the output writing? Maybe just `timer`?

File: src/lib/operators/join_sort_merge/radix_cluster_sort.hpp [hunk 5]
```diff
@@ -350,12 +360,15 @@ class RadixClusterSort {
...
       output.clusters_right = std::move(result.second);
     }
+    _performance.set_step_runtime(JoinSortMerge::OperatorSteps::Clustering, timer_clustering.lap());
 
     // Sort each cluster (right now std::sort -> but maybe can be replaced with
     // an more efficient algorithm if subparts are already sorted [InsertionSort?!])
     _sort_clusters(output.clusters_left);
     _sort_clusters(output.clusters_right);
 
+    _performance.set_step_runtime(JoinSortMerge::OperatorSteps::Sorting, timer_clustering.lap());
```
Comment (Bouncner) [inline src/lib/operators/join_sort_merge/radix_cluster_sort.hpp:370]: Same as above, naming.

**Review by Bouncner — APPROVED** [review]
LGTM.

**Comment by Bouncner** [comment]
@mrks: Do you want to have a look at this PR before merging?

**Comment by mrks** [comment]
Scrolled over it, lgtm.

**Pull request merged**